### PR TITLE
CDialogImpl for ATL

### PR DESCRIPTION
**Pull request by learn-more**
## Purpose

Implement CDialogImpl

JIRA issue: [CORE-13895](https://jira.reactos.org/browse/CORE-13895)

### CODE REVIEW

**Comment by yagoulas** [comment]
Looks fine after a quick view. I will give a more detailed view later.

File: sdk/lib/atl/atlwin.h [hunk 1]
```diff
@@ -1276,20 +1279,180 @@ class CWindow
...
+    }
+};
+
+
+template <class TBase = CWindow>
+class CDialogImplBaseT : public CWindowImplRoot<TBase>
+{
+public:
+    // + Hacks for gcc
+    using CWindowImplRoot<TBase>::WINSTATE_DESTROYED;
```
Comment (sanchaez) [inline sdk/lib/atl/atlwin.h]: This is caused by WINSTATE_DESTROYED being scoped unnamed enum and you use it as a global enum.Maybe it's better to name this enum: `enum WinState { ... };` and do ` using CWindowImplRoot<TBase>::WinState`?

Comment (learn-more) [reply sdk/lib/atl/atlwin.h]: No, it's defined in the base class, but gcc fails to resolve this due to the usage of templates.

Comment (sanchaez) [reply sdk/lib/atl/atlwin.h]: Well, you resolved it with `CWindowImplRoot<TBase>` part.

Comment (sanchaez) [reply sdk/lib/atl/atlwin.h]: It's not really GCC's fault, it's MSVC unqualified lookup extension/feature that allows it to compile. It is not part of the C++ Standard. 
You can also try `this->WINSTATE_DESTROYED`. It should work too.

File: sdk/lib/atl/atlwin.h [hunk 2]
```diff
@@ -1436,6 +1599,7 @@ class CWindowImplBaseT : public TBase, public CMessageMap
     }
 };
 
+
```
Comment (sanchaez) [inline sdk/lib/atl/atlwin.h]: Double newline.

File: sdk/lib/atl/atlwin.h [hunk 3]
```diff
@@ -1276,20 +1279,180 @@ class CWindow
...
+template <class T, class TBase = CWindow>
+class CDialogImpl : public CDialogImplBaseT< TBase >
+{
+public:
+    // + Hacks for gcc
+    using CWindowImplRoot<TBase>::m_thunk;
+    using CWindowImplRoot<TBase>::m_hWnd;
+    // - Hacks for gcc
+
+
```
Comment (sanchaez) [inline sdk/lib/atl/atlwin.h]: Double newline.

File: sdk/lib/atl/atlwin.h [hunk 4]
```diff
@@ -1276,20 +1279,180 @@ class CWindow
...
+        CDialogImplBaseT<TBase> *pThis;
+        DLGPROC newDlgProc;
+        DLGPROC GCCU(pOldProc);
+
+        pThis = reinterpret_cast<CDialogImplBaseT<TBase>*>(_AtlWinModule.ExtractCreateWndData());
+        ATLASSERT(pThis != NULL);
+        if (pThis == NULL)
+            return 0;
+        pThis->m_thunk.Init((WNDPROC)pThis->GetDialogProc(), pThis);
+        newDlgProc = (DLGPROC)pThis->m_thunk.GetWNDPROC();
```
Comment (sanchaez) [inline sdk/lib/atl/atlwin.h]: Should be `static_cast`? You use them elsewhere in this file.

File: sdk/lib/atl/atlwin.h [hunk 5]
```diff
@@ -1276,20 +1279,180 @@ class CWindow
...
+        HWND hWnd;
+        T* pImpl;
+
+        result = m_thunk.Init(NULL, NULL);
+        if (result == FALSE)
+            return NULL;
+
+        _AtlWinModule.AddCreateWndData(&m_thunk.cd, this);
+
+        pImpl = static_cast<T*>(this);
```
Comment (sanchaez) [inline sdk/lib/atl/atlwin.h]: ... here you use `static_cast`.

File: sdk/lib/atl/atlwin.h [hunk 6]
```diff
@@ -1276,20 +1279,180 @@ class CWindow
...
+        if (handled)
+        {
+            if ((pThis->m_dwState & WINSTATE_DESTROYED) == 0)
+            {
+                ::SetWindowLongPtr(pThis->m_hWnd, DWLP_MSGRESULT, lResult);
+            }
+        }
+        else
+        {
+            if (uMsg == WM_NCDESTROY)
```
Comment (sanchaez) [inline sdk/lib/atl/atlwin.h]: Maybe you can merge this into `else if (...) {`.

File: sdk/lib/atl/atlwin.h (hunk 4 above)
Comment (learn-more) [reply sdk/lib/atl/atlwin.h]: yup

File: sdk/lib/atl/atlwin.h (hunk 6 above)
Comment (learn-more) [reply sdk/lib/atl/atlwin.h]: i dont like that in this case

**Pull request merged**